support-models/src/main/scala/com/gu/support/workers/GiftRecipient.scala

```diff
@@ -1,12 +1,79 @@
 package com.gu.support.workers
 
+import cats.implicits._
 import com.gu.i18n.Title
 import com.gu.support.encoding.Codec
 import com.gu.support.encoding.Codec.deriveCodec
 import com.gu.support.encoding.CustomCodecs._
+import com.gu.support.workers.GiftRecipient.{DigiSub, Weekly}
+import io.circe._
+import io.circe.syntax._
 
-case class GiftRecipient(title: Option[Title], firstName: String, lastName: String, email: Option[String])
-
+sealed trait GiftRecipient {
+  val giftRecipientType: String
+  val firstName: String
+  val lastName: String
+  def asDigiSub: Option[DigiSub]
+  def asWeekly: Option[Weekly]
+}
 object GiftRecipient {
-  implicit val codec: Codec[GiftRecipient] = deriveCodec
+
+  implicit val c1: Codec[Weekly] = deriveCodec
+  implicit val c2: Codec[DigiSub] = deriveCodec
+
+  implicit val encoder: Encoder[GiftRecipient] = Encoder.instance {
+    case c: Weekly => c.asJson
+    case c: DigiSub => c.asJson
+  }
+
+  def or[A, AA >: A](da: => Decoder[A], daa: => Decoder[AA]): Decoder[AA] = new Decoder[AA] {
+    final def apply(c: HCursor): Decoder.Result[AA] = da(c) match {
+      case r @ Right(_) => r
+      case Left(failure)      => daa(c).leftMap(secondFail => DecodingFailure(failure.message + " / " + secondFail.message, failure.history ++ secondFail.history))
+    }
+  }
```

Comment:
the "or" method is not essential. It was needed for debugging, as the built in `or` tries them all, and throws away all but the last error.